### PR TITLE
Implements Session.destroy method which stops sending bosh requests and aborts the current requests

### PR DESCRIPTION
**Pull request by hristoterezov**
This method is almost the same as "end" except that it aborts the current requests and it isn't sending terminate packet. This should be useful for prebind.

### CODE REVIEW

File: lib/bosh.js [hunk 1]
```diff
@@ -161,6 +162,19 @@ BOSHConnection.prototype.end = function (stanzas) {
   }.bind(this))
 }
 
+BOSHConnection.prototype.destroy = function () {
+  delete this.sid
+  this.activeRequests.forEach(function (request) {
+    request.abort();
```
Comment (sonnyp) [inline lib/bosh.js]: no ;

File: lib/bosh.js [hunk 2]
```diff
@@ -180,15 +194,17 @@ BOSHConnection.prototype.request = function (attrs, children, cb, retry) {
...
     uri: this.boshURL,
     method: 'POST',
     headers: { 'Content-Type': this.contentType },
     body: boshEl.toString()
   },
     function (err, res, body) {
       that.currentRequests--
-
+      var currentRequestIndex = that.activeRequests.indexOf(currentRequest)
+      if(currentRequestIndex != -1)
```
Comment (sonnyp) [inline lib/bosh.js]: space before/after keywords

File: lib/bosh.js (hunk 1 above)
Comment (hristoterezov) [inline lib/bosh.js]: Fixed!

File: lib/bosh.js (hunk 2 above)
Comment (hristoterezov) [inline lib/bosh.js]: Fixed!

File: lib/bosh.js [hunk 3]
```diff
@@ -161,6 +162,19 @@ BOSHConnection.prototype.end = function (stanzas) {
...
+  this.activeRequests.forEach(function (request) {
+    request.abort()
+  })
+  this.activeRequests = []
+  this.shutdown = true
+  this.queue = []
+  this.emit('disconnect')
+  this.emit('end')
+  this.emit('close')
+}
```
Comment (sonnyp) [inline lib/bosh.js]: can you remove duplicated code from end and call destroy instead in the request callback ?

File: lib/bosh.js [hunk 4]
```diff
@@ -180,15 +194,18 @@ BOSHConnection.prototype.request = function (attrs, children, cb, retry) {
...
     headers: { 'Content-Type': this.contentType },
     body: boshEl.toString()
   },
     function (err, res, body) {
       that.currentRequests--
-
+      var currentRequestIndex = that.activeRequests.indexOf(currentRequest)
+      if (currentRequestIndex != -1) {
+        that.activeRequests.splice(currentRequestIndex, 1)
+      }
```
Comment (sonnyp) [inline lib/bosh.js]: please use a Map instead (use es6-collections module)

**Comment by sonnyp** [comment]
good idea, can you add tests please?

File: lib/bosh.js (hunk 3 above)
Comment (sonnyp) [inline lib/bosh.js]: actually in `end`, `destroy` should be called once all pending requests are done

Comment (sonnyp) [inline lib/bosh.js]: discard previous comments here

File: lib/bosh.js [hunk 5]
```diff
@@ -161,6 +162,19 @@ BOSHConnection.prototype.end = function (stanzas) {
   }.bind(this))
 }
 
+BOSHConnection.prototype.destroy = function () {
+  delete this.sid
+  this.activeRequests.forEach(function (request) {
+    request.abort()
+  })
+  this.activeRequests = []
```
Comment (sonnyp) [inline lib/bosh.js]: use a Map (see comment below) and do not reassign activeRequests, just delete them all

**Comment by sonnyp** [comment]
please resend pull request against https://github.com/node-xmpp/node-xmpp

node-xmpp-client moved there (but the npm package remain the same)

sorry for the inconvenience

File: lib/bosh.js (hunk 4 above)
Comment (hristoterezov) [inline lib/bosh.js]: @sonnyp Why do we need map for this? What should I put as keys?

**Comment by hristoterezov** [comment]
@sonnyp I'll resend the PR and I'll try to write some tests. Thanks for the comments and sorry for the delay.

**Comment by hristoterezov** [comment]
Moved to https://github.com/node-xmpp/node-xmpp/pull/320

**Pull request closed without merging**